### DISCUSSION TITLE
How to use insta-cmd with Crane

### DISCUSSION THREAD

**Discussion by msfjarvis** [ipetkov/crane#534] [Q&A] [answered]
[insta-cmd](https://github.com/mitsuhiko/insta-cmd) allows writing snapshot tests for CLI output which is great, but when building with Crane it fails to execute the tests because it tries to lookup the path to the crate's bin output in order to execute it and capture the output. I tried a bunch of things involving passing buildPhase commands only to realize later that the checkPhase runs in a clean directory anyway so I was on the wrong track.

**Comment by ipetkov** [comment]
Could you please share (an ideally minimal) reproduction of the issue?

I suspect that the issue might be in the `buildDepsOnly` derivation since we trim any package binary definitions, but I can't say for certain without more details

**Reply by msfjarvis** [reply]
Apologies, I should've included the link in the OP: https://github.com/msfjarvis/adx

The nextest derivation is currently disabled, you'll need to uncomment it: https://github.com/msfjarvis/adx/blob/ec0c50fb3d5df40cc971307ce8f05e983cecc7f9/flake.nix#L84

**Reply by ipetkov** [reply, marked as answer]
It looks like you have to build your actual binary before running the tests which require it:

```patch
diff --git a/flake.nix b/flake.nix
index 8df72eb..aed0d96 100644
--- a/flake.nix
+++ b/flake.nix
@@ -63,8 +63,14 @@
         cargoToml = ./adx/Cargo.toml;
       };
 
-      cargoArtifacts = craneLib.buildDepsOnly (commonArgs // {doCheck = false;});
-      adx = craneLib.buildPackage (commonArgs // {doCheck = false;});
+      cargoArtifacts = craneLib.buildDepsOnly commonArgs;
+      adx-build = craneLib.cargoBuild (commonArgs // {
+        inherit cargoArtifacts;
+      });
+      adx = craneLib.buildPackage (commonArgs // {
+        cargoArtifacts = adx-build;
+        doCheck = false;
+      });
       adx-clippy = craneLib.cargoClippy (commonArgs
         // {
           inherit cargoArtifacts;
@@ -73,7 +79,7 @@
       adx-audit = craneLib.cargoAudit (commonArgs // {inherit advisory-db;});
       adx-nextest = craneLib.cargoNextest (commonArgs
         // {
-          inherit cargoArtifacts;
+          cargoArtifacts = adx-build;
           partitions = 1;
           partitionType = "count";
         });
@@ -81,7 +87,7 @@
       checks = {
         inherit adx adx-clippy adx-fmt;
         # TODO: figure out how to support insta-cmd here
-        # inherit adx-nextest;
+        inherit adx-nextest;
         # TODO: Re-enable once https://github.com/NixOS/nixpkgs/issues/288064 is fixed
         # inherit adx-audit;
       };
```

**Reply by msfjarvis** [reply]
That did the trick, thank you! I wasn't aware of `cargoBuild` :)
